File: src/dataset/c/beer_song_Exercism_vlzware/claude/ClaudeSonnet4_beer_song_v3.c
#include "beer_song.h"
#include <stdio.h>
#include <string.h>

static const char bottle[] = " bottle";
static const char bottles[] = " bottles";
static const char on_the_wall[] = " of beer on the wall";
static const char of_beer[] = " of beer.\n";
static const char take_one[] = "Take one down and pass it around, ";
static const char take_it[] = "Take it down and pass it around, ";
static const char no_more[] = "No more";
static const char no_more_n[] = "no more";
static const char go_to_store[] = "Go to the store and buy some more, ";

void mcat(int *pos, char *dst, const char *src_);

void verse(char *buf, int count)
{
    if (count > 99 || count < 0) {
        fprintf(stderr, "Invalid bottle count.\n");
        return;
    }

    char *ptr = buf;
    char num_str[4];
    
    if (count == 0) {
        strcpy(ptr, no_more);
        ptr += sizeof(no_more) - 1;
    } else {
        int len = sprintf(num_str, "%d", count);
        memcpy(ptr, num_str, len);
        ptr += len;
    }
    
    const char *bottle_word = (count == 1) ? bottle : bottles;
    strcpy(ptr, bottle_word);
    ptr += strlen(bottle_word);
    
    strcpy(ptr, on_the_wall);
    ptr += sizeof(on_the_wall) - 1;
    
    strcpy(ptr, ", ");
    ptr += 2;
    
    if (count == 0) {
        strcpy(ptr, no_more_n);
        ptr += sizeof(no_more_n) - 1;
    } else {
        memcpy(ptr, num_str, strlen(num_str));
        ptr += strlen(num_str);
    }
    
    strcpy(ptr, bottle_word);
    ptr += strlen(bottle_word);
    
    strcpy(ptr, of_beer);
    ptr += sizeof(of_beer) - 1;
    
    if (count == 0) {
        strcpy(ptr, go_to_store);
        ptr += sizeof(go_to_store) - 1;
    } else if (count == 1) {
        strcpy(ptr, take_it);
        ptr += sizeof(take_it) - 1;
    } else {
        strcpy(ptr, take_one);
        ptr += sizeof(take_one) - 1;
    }
    
    int next_count = (count == 0) ? 99 : count - 1;
    if (count == 1) {
        strcpy(ptr, no_more_n);
        ptr += sizeof(no_more_n) - 1;
    } else {
        int len = sprintf(num_str, "%d", next_count);
        memcpy(ptr, num_str, len);
        ptr += len;
    }
    
    bottle_word = (next_count == 1) ? bottle : bottles;
    strcpy(ptr, bottle_word);
    ptr += strlen(bottle_word);
    
    strcpy(ptr, on_the_wall);
    ptr += sizeof(on_the_wall) - 1;
    
    strcpy(ptr, ".\n");
    ptr += 2;
    
    *ptr = '\0';
}

void sing(char *buf, int from, int to)
{
    if (to > from)
        return;
        
    char *ptr = buf;
    char verse_buf[512];
    
    for (int i = from; i >= to; i--) {
        verse(verse_buf, i);
        size_t verse_len = strlen(verse_buf);
        memcpy(ptr, verse_buf, verse_len);
        ptr += verse_len;
        
        if (i > to) {
            *ptr++ = '\n';
        }
    }
    
    *ptr = '\0';
}

void mcat(int *pos, char *dst, const char *src_)
{
    const char *src = src_;
    char *tmp = dst + *pos;
    
    while (*src) {
        *tmp++ = *src++;
        (*pos)++;
    }
    
    *tmp = '\0';
}

File: src/dataset/Java/beer_song_Exercism_java_rabestro/claude/ClaudeSonnet4_BeerSong_v3.java
class BeerSong {

    private static final int INITIAL_BOTTLES = 99;
    private static final String[] BOTTLE_CACHE = new String[101];
    private static final String NO_MORE_BOTTLES = "no more bottles";
    private static final String ONE_BOTTLE = "1 bottle";
    private static final String BOTTLES_SUFFIX = " bottles";
    private static final String BEER_ON_WALL = " of beer on the wall, ";
    private static final String BEER_NEWLINE = " of beer.\n";
    private static final String TAKE_ONE = "Take one down and pass it around, ";
    private static final String TAKE_IT = "Take it down and pass it around, ";
    private static final String BEER_ON_WALL_END = " of beer on the wall.\n\n";
    private static final String NO_MORE_START = "No more bottles of beer on the wall, no more bottles of beer.\nGo to the store and buy some more, ";

    static {
        BOTTLE_CACHE[0] = NO_MORE_BOTTLES;
        BOTTLE_CACHE[1] = ONE_BOTTLE;
        for (int i = 2; i <= 100; i++) {
            BOTTLE_CACHE[i] = i + BOTTLES_SUFFIX;
        }
    }

    public String sing(int count, int repeat) {
        final var song = new StringBuilder(repeat * 150);
        while (repeat-- > 0) {
            if (count == 0) {
                song.append(NO_MORE_START);
                count = INITIAL_BOTTLES;
            } else {
                String currentBottle = BOTTLE_CACHE[count];
                song.append(currentBottle).append(BEER_ON_WALL)
                    .append(currentBottle).append(BEER_NEWLINE)
                    .append(count == 1 ? TAKE_IT : TAKE_ONE);
                --count;
            }
            song.append(BOTTLE_CACHE[count]).append(BEER_ON_WALL_END);
        }
        return song.toString();
    }

    public String singSong() {
        return sing(INITIAL_BOTTLES, 100);
    }

    private String bottle(int count) {
        return BOTTLE_CACHE[count];
    }
}